Fix Site Editor perf tests

WunderBart opened the pull request:
## What?

Isolating the solution from https://github.com/WordPress/gutenberg/pull/48208. See https://github.com/WordPress/gutenberg/pull/48208#issuecomment-1436030942.

## Why?

Site Editor performance tests are currently very flaky.

## How?

Focus on the paragraph instead of clicking it.

kevin940726 committed:
Focus on the paragraph instead clicking it
See https://github.com/WordPress/gutenberg/pull/48208#issuecomment-1436030942

kevin940726 commented:
See https://github.com/WordPress/gutenberg/pull/48208#issuecomment-1436014572, [9bc7099](https://github.com/WordPress/gutenberg/pull/48208/commits/9bc7099358989729d80340a7b1ee3f48ae2f59d1) is also required to fix another failure (there are two issues). Since it was introduced in #48063, @youknowriad might have an idea why it solves the issue and what the solution should be.

kevin940726 commented on packages/e2e-tests/specs/performance/site-editor.test.js:
```diff
@@ -147,8 +147,9 @@ describe( 'Site Editor Performance', () => {
-			'[data-type="core/post-content"] [data-type="core/paragraph"]'
+		await canvas().focus(
+			'[data-type="core/post-content"] [data-type="core/paragraph"]',
+			{ timeout: 240000 }
```
I don't think we need this timeout anymore 🙂 .

t-hamano commented:
To genuinely solve this problem, the following additional changes would need to be made.

https://github.com/WordPress/gutenberg/pull/48208/files#diff-686c21b911380ccd4d7165b856981854a1dca8bf2084e4e4a5e487219e6afc63R179-R182
https://github.com/WordPress/gutenberg/pull/48208/files#diff-dac6b5792388c0d2844f5ec7201465c2a1d35b2bc18c580e0aa811a812af2f17R62
https://github.com/WordPress/gutenberg/pull/48208/files#diff-dac6b5792388c0d2844f5ec7201465c2a1d35b2bc18c580e0aa811a812af2f17R74
https://github.com/WordPress/gutenberg/pull/48208/files#diff-dac6b5792388c0d2844f5ec7201465c2a1d35b2bc18c580e0aa811a812af2f17R93

@kevin940726 
If you see something that does not need to be changed, please let us know 🙏

WunderBart committed:
Add missing fixes
As per https://github.com/WordPress/gutenberg/pull/48240#pullrequestreview-1305477370 and https://github.com/WordPress/gutenberg/pull/48240#issuecomment-1436701713

WunderBart commented:
Missing bits added as per https://github.com/WordPress/gutenberg/pull/48240#issuecomment-1436701713 and https://github.com/WordPress/gutenberg/pull/48240#pullrequestreview-1305477370 🤞

youknowriad commented:
@t-hamano The last link shouldn't be needed after https://github.com/WordPress/gutenberg/pull/48063

youknowriad commented on packages/e2e-tests/specs/performance/site-editor.test.js:
```diff
@@ -90,6 +90,7 @@ describe( 'Site Editor Performance', () => {
 		await visitSiteEditor( {
 			postId: id,
 			postType: 'page',
+			path: '/navigation/single',
```
This is not needed I think.

t-hamano replied:
Somehow, this was necessary for the E2E test. See [this comment](https://github.com/WordPress/gutenberg/pull/48208#issuecomment-1435988885) and [this comment](https://github.com/WordPress/gutenberg/pull/48208#issuecomment-1436014572).

WunderBart replied:
I've reverted it in https://github.com/WordPress/gutenberg/pull/48240/commits/87b1109c0517b27ffe8f5bd5fb417a37e7a822d4 since the specs failed without it.

WunderBart replied:
Any idea why it would fail in CI but not locally?

youknowriad replied:
Maybe you started the PR before my refactoring PR was merged. What if you rebase?

youknowriad replied:
That path is not used anymore (you can search the code base) so I have no other explanation that this is just a random failure or is using outdated code.

WunderBart replied:
> Maybe you started the PR before my refactoring PR was merged. What if you rebase?

This branch was created from a recent trunk (last commit is https://github.com/WordPress/gutenberg/commit/8f9c3165c5131827fdd8c73a6c1a51d29c0aeb55), so I don't think it would be the case.

WunderBart replied:
The tests have passed with this addition. I'll rerun 2-3 more times to be sure it's not a fluke.

_edit: Just saw https://github.com/WordPress/gutenberg/pull/48240#discussion_r1111856946 - it's likely a fluke then._

WunderBart replied:
@youknowriad, it actually passed 3 times in a row with this exact addition. I'll remove the path param and see if it still passes.

WunderBart replied:
It [failed](https://github.com/WordPress/gutenberg/actions/runs/4223844027/jobs/7334052585). [Re-adding](https://github.com/WordPress/gutenberg/pull/48240/commits/da697d8398dd4ff7340bd4e22733eee86b23c747) to be extra sure it fixes the issue.

youknowriad replied:
Weird, I'm fine keeping it for now but we'll have to figure out the reason because thet route don't exist anymore. Maybe the presence of the "path" argument triggers some condition.

t-hamano replied:
I am not sure I understand it in depth but is this line relevant?
https://github.com/WordPress/gutenberg/blob/aa1fca2c794a78b5153de07c5d6c8863f5c81795/packages/edit-site/src/components/sync-state-with-url/use-sync-path-with-url.js#L67

youknowriad replied:
yep but since there's a postId and postType, we're supposed to enter the condition right after it which basically overrides the path

t-hamano replied:
I found a thing of interest. In **Firefox**, accessing the following URL will cause a redirect and the page will not be displayed correctly.

From:
```
http://localhost:8888/wp-admin/site-editor.php?canvas=view&postType=page&postId=2
```

Redirected to:
```
http://localhost:8888/wp-admin/site-editor.php?postId=none
```

![firefox](https://user-images.githubusercontent.com/54422211/220130147-435d5f4c-cf60-4214-ac02-64cfe60a8e84.png)

youknowriad replied:
Good find. 

So we do have [this code](https://github.com/WordPress/gutenberg/blob/aa1fca2c794a78b5153de07c5d6c8863f5c81795/lib/compat/wordpress-6.2/site-editor.php#L34-L45 ) in the Gutenberg plugin that forces the site editor to redirect to "postId:none". But the reasoning for that code is that it's a filter for `wp_redirect` meaning it only runs if we're redirecting to the site editor. Something we do in WP core (in 6.1 and before) but that redirect is not supposed to happen if we provide a postId and postType in the URL https://github.com/WordPress/wordpress-develop/blob/73f7aa4a149890f2e565ac74d9d49b5258b8b984/src/wp-admin/site-editor.php#L36-L46

So the question here is why the redirect filter is triggering while we do have postId and postType in the URL.

WunderBart replied:
Tests passed! 🥳 I need to head out now, so feel free to ship.

kevin940726 replied:
Okay, I found out why this is happening on CI. I think it's because it doesn't have some required cookies, yet. You can reproduce it locally in incognito mode. Hope that it will spark some ideas for the solution.

Meanwhile, I think we should just merge this to unblock other PRs. We can figure out why and the proper solution in another PR. WDYT?

t-hamano replied:
How about trying to remove the code in [this comment](https://github.com/WordPress/gutenberg/pull/48240#discussion_r1111824772) before merging?

t-hamano replied:
@youknowriad 

> I found a thing of interest. In Firefox, accessing the following URL will cause a redirect and the page will not be displayed correctly.

Sorry, this was my mistake. I was simply not logged into WordPress 😅 Therefore, this problem also occurs in Chrome. This issue may need to be looked into separately, as this URL should originally redirect to the login page.

youknowriad commented on packages/e2e-tests/specs/performance/site-editor.test.js:
```diff
@@ -71,7 +71,7 @@ describe( 'Site Editor Performance', () => {
 			} );
 
-			dispatch( 'core/block-editor' ).resetBlocks( blocks );
+			await dispatch( 'core/block-editor' ).resetBlocks( blocks );
```
This change is a bit weird for me. AFAIK, this action is synchronous

kevin940726 replied:
`resetBlocks()` returns a promise though. I think it's because it is a thunk: https://github.com/WordPress/gutenberg/blob/a770a6eaceacab0ae499950b4c539676662f4148/packages/block-editor/src/store/actions.js#L68-L73

But this is likely not the root cause of the failure. We can omit this in this PR.

WunderBart committed:
Try without the extra path param
As per https://github.com/WordPress/gutenberg/pull/48240#pullrequestreview-1305501786

WunderBart committed:
Remove obsolete await
As per https://github.com/WordPress/gutenberg/pull/48240#discussion_r1111761678

WunderBart committed:
Revert "Try without the extra path param"

This reverts commit 5fad8cd10c0765b8054ffdae7f6bd1840c7bba6c.

kevin940726 commented on packages/e2e-test-utils/src/site-editor.js:
```diff
@@ -175,4 +175,9 @@ export async function enterEditMode() {
 	}
 	await canvas().click( 'body' );
+
+	await page.waitForSelector(
+		'[role="region"][aria-label="Navigation sidebar"]',
+		{ hidden: true }
+	);
```
This is probably not needed. Let's try removing it?

WunderBart committed:
Revert "Revert "Try without the extra path param""

This reverts commit 87b1109c0517b27ffe8f5bd5fb417a37e7a822d4.

WunderBart committed:
Revert "Revert "Revert "Try without the extra path param"""

This reverts commit 2b931a8bccfbafa9665e638045c6d1847955d98e.

ellatrix commented on packages/e2e-tests/specs/performance/site-editor.test.js:
```diff
@@ -147,7 +148,7 @@ describe( 'Site Editor Performance', () => {
 		);
 		await enterEditMode();
-		await canvas().click(
+		await canvas().focus(
```
Since the `path` arg is the problem, is this change still needed?

kevin940726 replied:
Yes, it's needed. It's mentioned above in https://github.com/WordPress/gutenberg/pull/48240#issuecomment-1436691713, in the PR description, and in the [original exploration PR](https://github.com/WordPress/gutenberg/pull/48208#issuecomment-1436014572), that there are two separate issues in the perf tests. The `path` one is always failing, while this is a really flaky one. I think we need both to unblock others.

ellatrix committed:
Remove unneeded changes

dmsnell committed:
Describe discrepancy of inexplicably-required line in code via explanatory comment.

dmsnell reviewed (approved):
I've added a comment pointing to the discussion here concerning `/navigation/single`.

For now if it's clears the tests then I encourage a merge to stop blocking unrelated work. In #48262 I've made a crude attempt to revert the new-page-per-e2e-test and see if that's related to these failures (may not have properly applied the revert), though the merge of that PR was ten days ago and the failures seemed to have started closer to four days ago.

t-hamano committed:
Use `canvas.focus()` instead of `canvas.click()`

t-hamano commented:
I have re-applied the changes mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/48240#discussion_r1112377179). Let't try to see if the test passes several times.

t-hamano commented:
The performance test was [successful three times](https://github.com/WordPress/gutenberg/actions/runs/4228613161/jobs/7344893239). Let's merge!

ntsekouras commented:
I just cherry-picked this PR to the wp/6.2 branch.
